library/std/src/sync/once_lock.rs

```diff
@@ -14,7 +14,10 @@ use crate::sync::Once;
 /// # Examples
 ///
 /// Using `OnceCell` to store a function’s previously computed value (a.k.a.
-/// ‘lazy static’ or ‘memoizing’):
+/// ‘lazy static’ or ‘memoizing’).  Since [`drop`] is not called for static
+/// items at the end of the program (see [the
+/// reference](https://doc.rust-lang.org/reference/destructors.html), objects
+/// stored using this pattern will not have their destructors called.
```

Comment:
I'm not necessarily against adding this somewhere in the documentation but why to OnceLock's example, of all places?

Reply:
AFAICT, the primary purpose of OnceLock is to manage a static item, and at least to me it wasn't obvious that a consequence of moving an item in to a (static) OnceLock that is that the item's destructor won't be called.

Reply:
The codebase I work with has more uses of OnceCell and OnceLock in fields of random structs than in statics. A lot of statics are used to provide global mutable state which accumulates over time, which often requires something like Mutex or a thread_local RefCell.

Notably, we do mention this problem in [the documentation of the static keyword](https://doc.rust-lang.org/std/keyword.static.html).

Reply:
Interesting - I'm surprised to learn that OnceLock is commonly used in fields of random non-static structs, would be keen to see an example of that usage in the docs.

FWIW, my "systems" programming background is largely in C++, which does destruct statics.  To me, OnceLock looks simply like a Rust implementation of a global singleton, and it was surprising that putting a struct in the idiomatic (or, at least following the example from docs) resulted in this behaviour where that object isn't dropped.

I appreciate that the docs already explain that statics aren't dropped; the point of the PR is to indicate that the docs around drop and static might be interesting to someone who's applying the example.  IME static isn't a very widely used keyword, so a person who's looking at the OnceLock example probably isn't super familiar with the ramifications of using it.

Would a comment in the example code perhaps be a better place for this note?

Reply:
If the entire struct has to be Sync for other reasons, it becomes natural to use OnceLock, because OnceCell is !Sync.

I'd be happy to see a comment in the examples, yeah, or even... hmm.

See, _most_ of the Drops in std are "and this gets deallocated" (Vec) or "this gets advanced towards deallocation" (Arc), which means that once the program ends, it doesn't really matter: the OS sends all the types to the Big Drop Impl In The Sky. So I think this might be more motivated by a (perhaps slightly contrived?) example which presents something with a slightly-complicated Drop interacting with OnceCell/OnceLock in different ways. I suspect you encountered some sort of stumbling block that deserves more than just a sentence.

Reply:
The issue I encountered relates to a proprietary and rather crude Linux kernel-mode driver; basically the library I'm working on should make a call in to that driver to release a hardware resource, and that was done by destructing a Rust type.  It's OK if abnormal program termination doesn't call drop on this type, because the "abnormal" aspect means we need to reset the whole system anyway, but it's problematic if normal termination doesn't release that resource.

But on a more general note, as before I found this surprising and bounced it off a couple other Rust folks (including [in Discord](https://discord.com/channels/273534239310479360/273541522815713281/1179973533082849400)) who didn't seem to know the right answer offhand either.

Will aim to update this PR later today with a code comment instead, thanks for your feedback!

Reply:
Hm. I'm sorry, "abnormal program termination" doesn't actually mean anything to me, I'm a complete feral by systems programmer standards. I've seen people use it to refer specifically to `abort()`, but consider `std::process::exit(exit_code)` to be normal termination. I'm concerned about that statement, because that _also_ will not call destructors.

Reply:
Sorry, I didn't mean anything technical/specific by "abnormal program termination".

If we imagine a spectrum of reasons why a program ended, with the extreme normal end "program finished its job", and "CPU exploded" as the extreme abnormal end; in this case `MyStruct::drop()` needs to be called only if the program finished its job.  Anywhere else on the spectrum, the system is restarted anyway.

The issue here is that, by moving MyStruct in to the idiomatic (per docs) OnceLock usage, I have wound up with its drop never being called.